File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Benchmarks/ViewModels/BenchmarkViewModel.swift
//
//  BenchmarkViewModel.swift
//  RunAnywhereAI
//
//  Orchestrates benchmark execution, persistence, and export.
//

import Foundation
import RunAnywhere
import os
#if canImport(UIKit)
import UIKit
#endif

@MainActor
@Observable
final class BenchmarkViewModel {
    /// Grep marker for TC-19 harness (`test_workflows/scripts/_log_markers.sh`).
    private static let historySavedLogMarker = "Benchmark history saved"
    // MARK: - State

    var isRunning = false
    var progress: Double = 0
    var currentScenario: String = ""
    var currentModel: String = ""
    var completedCount: Int = 0
    var totalCount: Int = 0
    var pastRuns: [BenchmarkRun] = []
    var selectedCategories: Set<BenchmarkCategory> = Set(BenchmarkCategory.allCases)
    var errorMessage: String?
    var showClearConfirmation = false

    /// Brief toast message shown after clipboard copy
    var copiedToastMessage: String?

    /// Categories that had no downloaded models during the last run
    var skippedCategoriesMessage: String?

    /// Available downloaded models grouped by category, for user selection
    var availableModels: [BenchmarkCategory: [RAModelInfo]] = [:]

    /// Which model IDs the user has selected for benchmarking.
    /// Starts empty and is populated from the downloaded-model registry.
    var selectedModelIds: Set<String> = []

    // MARK: - Private

    private let runner = BenchmarkRunner()
    private let store = BenchmarkStore()
    private let logger = Logger(subsystem: "com.runanywhere.RunAnywhereAI", category: "Benchmarks")
    private var runTask: Task<Void, Never>?
    private var toastTask: Task<Void, Never>?

    // MARK: - Lifecycle

    func loadPastRuns() {
        pastRuns = store.loadRuns().reversed()
    }

    func refreshAvailableModels() {
        Task {
            await reloadAvailableModels()
        }
    }

    /// Resync registry paths from disk, then rebuild the grouped model picker state.
    private func reloadAvailableModels() async {
        await RunAnywhere.refreshModelRegistry()

        let listResult = await RunAnywhere.listModels()
        guard listResult.success else {
            availableModels = [:]
            return
        }
        let allModels = listResult.models.models
        var grouped: [BenchmarkCategory: [RAModelInfo]] = [:]
        for category in BenchmarkCategory.allCases {
            let models = BenchmarkRunner.downloadedModels(for: category, in: allModels)
            if !models.isEmpty {
                grouped[category] = models
            }
        }
        availableModels = grouped
        let availableIds = Set(grouped.values.flatMap { $0 }.map { $0.id })
        if selectedModelIds.isEmpty {
            selectedModelIds = availableIds
        } else {
            selectedModelIds = selectedModelIds.intersection(availableIds)
            if selectedModelIds.isEmpty, !availableIds.isEmpty {
                selectedModelIds = availableIds
            }
        }
    }

    /// Categories that have on-disk models after the latest registry rescan.
    private func categoriesReadyToRun() -> Set<BenchmarkCategory> {
        let withModels = selectedCategories.filter { !(availableModels[$0]?.isEmpty ?? true) }
        return withModels.isEmpty ? selectedCategories : withModels
    }

    func toggleModel(_ modelId: String) {
        if selectedModelIds.contains(modelId) {
            selectedModelIds.remove(modelId)
        } else {
            selectedModelIds.insert(modelId)
        }
    }

    func selectAllModels() {
        selectedModelIds = Set(availableModels.values.flatMap { $0 }.map { $0.id })
    }

    func deselectAllModels() {
        selectedModelIds.removeAll()
    }

    // MARK: - Run

    func runBenchmarks() {
        guard !isRunning else { return }
        isRunning = true
        errorMessage = nil
        skippedCategoriesMessage = nil
        progress = 0
        completedCount = 0
        totalCount = 0
        currentScenario = "Preparing..."
        currentModel = ""

        runTask = Task {
            await executeBenchmarkRun()
        }
    }

    private func executeBenchmarkRun() async {
        await reloadAvailableModels()

        let availableIds = Set(availableModels.values.flatMap { $0 }.map { $0.id })
        if !availableIds.isEmpty {
            selectedModelIds = availableIds
        }
        let categoriesToRun = categoriesReadyToRun()

        let deviceInfo = makeDeviceInfo()
        var run = BenchmarkRun(deviceInfo: deviceInfo)

        do {
            let modelIds: Set<String>? = availableIds.isEmpty ? nil : availableIds
            let output = try await runner.runBenchmarks(
                categories: categoriesToRun,
                modelIds: modelIds
            ) { [weak self] update in
                Task { @MainActor in
                    self?.progress = update.progress
                    self?.completedCount = update.completedCount
                    self?.totalCount = update.totalCount
                    self?.currentScenario = update.currentScenario
                    self?.currentModel = update.currentModel
                }
            }

            if !output.skippedCategories.isEmpty {
                let names = output.skippedCategories.map(\.displayName).joined(separator: ", ")
                skippedCategoriesMessage = "Skipped (no models): \(names)"
            }

            run.results = output.results
            run.status = output.results.allSatisfy(\.metrics.didSucceed) ? .completed : .failed
            run.completedAt = Date()
        } catch is CancellationError {
            run.status = .cancelled
            run.completedAt = Date()
        } catch let error as BenchmarkRunnerError {
            run.status = .failed
            run.completedAt = Date()
            errorMessage = error.localizedDescription
        } catch {
            run.status = .failed
            run.completedAt = Date()
            errorMessage = error.localizedDescription
        }

        persistRunIfNeeded(run)
        loadPastRuns()
        isRunning = false
    }

    private func persistRunIfNeeded(_ run: BenchmarkRun) {
        // Persist completed work only — skip empty failed preflight runs so TC-19
        // grading waits for a real benchmark history entry.
        guard !run.results.isEmpty || run.status == .cancelled else { return }
        store.save(run: run)
        guard !run.results.isEmpty else { return }
        let duration = run.duration ?? 0
        let marker = Self.historySavedLogMarker
        logger.info(
            "\(marker, privacy: .public) results=\(run.results.count) duration=\(duration, privacy: .public)s"
        )
    }

    func cancel() {
        runTask?.cancel()
        runTask = nil
    }

    func clearAllResults() {
        store.clearAll()
        pastRuns = []
    }

    // MARK: - Copy to Clipboard

    func copyToClipboard(run: BenchmarkRun, format: BenchmarkExportFormat) {
        #if canImport(UIKit)
        let report = BenchmarkReportFormatter.formattedString(run: run, format: format)
        UIPasteboard.general.string = report
        let generator = UINotificationFeedbackGenerator()
        generator.notificationOccurred(.success)
        copiedToastMessage = "\(format.displayName) copied!"
        // Auto-dismiss after 2s — cancel any previous dismiss task
        toastTask?.cancel()
        toastTask = Task {
            try? await Task.sleep(for: .seconds(2))
            copiedToastMessage = nil
        }
        #endif
    }

    // MARK: - File Export

    func shareJSON(run: BenchmarkRun) -> URL {
        BenchmarkReportFormatter.writeJSON(run: run)
    }

    func shareCSV(run: BenchmarkRun) -> URL {
        BenchmarkReportFormatter.writeCSV(run: run)
    }

    // MARK: - Helpers

    private func makeDeviceInfo() -> BenchmarkDeviceInfo {
        if let sysInfo = DeviceInfoService.shared.deviceInfo {
            return BenchmarkDeviceInfo.fromSystem(sysInfo)
        }
        return BenchmarkDeviceInfo(
            modelName: "Unknown",
            chipName: "Unknown",
            totalMemoryBytes: Int64(ProcessInfo.processInfo.physicalMemory),
            availableMemoryBytes: SyntheticInputGenerator.availableMemoryBytes(),
            osVersion: ProcessInfo.processInfo.operatingSystemVersionString
        )
    }
}

File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/benchmarks/benchmark_view_model.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:runanywhere/runanywhere.dart' as sdk;
import 'package:runanywhere_ai/core/services/device_info_service.dart';
import 'package:runanywhere_ai/features/benchmarks/benchmark_runner.dart';
import 'package:runanywhere_ai/features/benchmarks/benchmark_store.dart';
import 'package:runanywhere_ai/features/benchmarks/benchmark_types.dart';

/// Orchestrates benchmark execution, persistence, and history.
///
/// Mirrors iOS `BenchmarkViewModel.swift` as a [ChangeNotifier] (precedent:
/// `model_list_view_model.dart`). Export/clipboard features are an iOS-only
/// extra and are omitted at MVP.
class BenchmarkViewModel extends ChangeNotifier {
  // MARK: - State

  bool isRunning = false;
  double progress = 0;
  String currentScenario = '';
  String currentModel = '';
  int completedCount = 0;
  int totalCount = 0;
  List<BenchmarkRun> pastRuns = [];
  Set<BenchmarkCategory> selectedCategories =
      BenchmarkCategory.values.toSet();
  String? errorMessage;

  /// Categories that had no downloaded models during the last run.
  String? skippedCategoriesMessage;

  /// Available downloaded models grouped by category, for user selection.
  Map<BenchmarkCategory, List<sdk.ModelInfo>> availableModels = {};

  /// Which model IDs the user has selected for benchmarking.
  Set<String> selectedModelIds = {};

  // MARK: - Private

  final BenchmarkRunner _runner = BenchmarkRunner();
  final BenchmarkStore _store = BenchmarkStore();
  bool _cancelRequested = false;

  // MARK: - Lifecycle

  Future<void> loadPastRuns() async {
    pastRuns = (await _store.loadRuns()).reversed.toList();
    notifyListeners();
  }

  void refreshAvailableModels() {
    unawaited(_reloadAvailableModels().then((_) => notifyListeners()));
  }

  /// Resync registry paths from disk, then rebuild the grouped model picker
  /// state (mirrors iOS `reloadAvailableModels`).
  Future<void> _reloadAvailableModels() async {
    try {
      await sdk.RunAnywhere.models.refreshModelRegistry();
      final listResult = await sdk.RunAnywhere.models.list();
      if (!listResult.success) {
        availableModels = {};
        return;
      }
      final allModels = listResult.models.models;
      final grouped = <BenchmarkCategory, List<sdk.ModelInfo>>{};
      for (final category in BenchmarkCategory.values) {
        final models = BenchmarkRunner.downloadedModels(category, allModels);
        if (models.isNotEmpty) {
          grouped[category] = models;
        }
      }
      availableModels = grouped;
    } catch (e) {
      debugPrint('Benchmarks: failed to reload models: $e');
      availableModels = {};
      return;
    }

    final availableIds = availableModels.values
        .expand((models) => models)
        .map((model) => model.id)
        .toSet();
    if (selectedModelIds.isEmpty) {
      selectedModelIds = availableIds;
    } else {
      selectedModelIds = selectedModelIds.intersection(availableIds);
      if (selectedModelIds.isEmpty && availableIds.isNotEmpty) {
        selectedModelIds = availableIds;
      }
    }
  }

  /// Categories that have on-disk models after the latest registry rescan.
  Set<BenchmarkCategory> _categoriesReadyToRun() {
    final withModels = selectedCategories
        .where((category) => availableModels[category]?.isNotEmpty ?? false)
        .toSet();
    return withModels.isEmpty ? selectedCategories : withModels;
  }

  void toggleCategory(BenchmarkCategory category) {
    if (selectedCategories.contains(category)) {
      selectedCategories.remove(category);
    } else {
      selectedCategories.add(category);
    }
    notifyListeners();
  }

  void toggleModel(String modelId) {
    if (selectedModelIds.contains(modelId)) {
      selectedModelIds.remove(modelId);
    } else {
      selectedModelIds.add(modelId);
    }
    notifyListeners();
  }

  void selectModels(Iterable<String> modelIds) {
    selectedModelIds.addAll(modelIds);
    notifyListeners();
  }

  void deselectModels(Iterable<String> modelIds) {
    selectedModelIds.removeAll(modelIds);
    notifyListeners();
  }

  void dismissError() {
    errorMessage = null;
    notifyListeners();
  }

  // MARK: - Run

  void runBenchmarks({bool allCategories = false}) {
    if (isRunning) return;
    if (allCategories) {
      selectedCategories = BenchmarkCategory.values.toSet();
    }
    isRunning = true;
    errorMessage = null;
    skippedCategoriesMessage = null;
    progress = 0;
    completedCount = 0;
    totalCount = 0;
    currentScenario = 'Preparing...';
    currentModel = '';
    _cancelRequested = false;
    notifyListeners();

    unawaited(_executeBenchmarkRun());
  }

  Future<void> _executeBenchmarkRun() async {
    await _reloadAvailableModels();

    final availableIds = availableModels.values
        .expand((models) => models)
        .map((model) => model.id)
        .toSet();
    if (availableIds.isNotEmpty) {
      selectedModelIds = availableIds;
    }
    final categoriesToRun = _categoriesReadyToRun();

    final run = BenchmarkRun(deviceInfo: _makeDeviceInfo());

    try {
      final output = await _runner.runBenchmarks(
        categories: categoriesToRun,
        modelIds: availableIds.isEmpty ? null : availableIds,
        isCancelled: () => _cancelRequested,
        onProgress: (update) {
          progress = update.progress;
          completedCount = update.completedCount;
          totalCount = update.totalCount;
          currentScenario = update.currentScenario;
          currentModel = update.currentModel;
          notifyListeners();
        },
      );

      if (output.skippedCategories.isNotEmpty) {
        final names = output.skippedCategories
            .map((category) => category.displayName)
            .join(', ');
        skippedCategoriesMessage = 'Skipped (no models): $names';
      }

      run.results = output.results;
      run.status = output.results
              .every((result) => result.metrics.didSucceed)
          ? BenchmarkRunStatus.completed
          : BenchmarkRunStatus.failed;
      run.completedAt = DateTime.now();
    } on BenchmarkCancelled {
      run.status = BenchmarkRunStatus.cancelled;
      run.completedAt = DateTime.now();
    } catch (e) {
      run.status = BenchmarkRunStatus.failed;
      run.completedAt = DateTime.now();
      errorMessage = e.toString();
    }

    // Persist completed work only — skip empty failed preflight runs
    // (mirrors iOS persistRunIfNeeded).
    if (run.results.isNotEmpty || run.status == BenchmarkRunStatus.cancelled) {
      await _store.save(run);
    }
    await loadPastRuns();
    isRunning = false;
    notifyListeners();
  }

  void cancel() {
    _cancelRequested = true;
  }

  Future<void> clearAllResults() async {
    await _store.clearAll();
    pastRuns = [];
    notifyListeners();
  }

  // MARK: - Helpers

  BenchmarkDeviceInfo _makeDeviceInfo() {
    final info = DeviceInfoService.shared.deviceInfo;
    return BenchmarkDeviceInfo(
      modelName: info?.modelName ?? 'Unknown',
      chipName: info?.chipName ?? 'Unknown',
      totalMemoryBytes: info?.totalMemory ?? 0,
      availableMemoryBytes: info?.availableMemory ?? 0,
      osVersion: info?.osVersion ?? 'Unknown',
    );
  }
}
